File: src/algorithms/python/flood-fill.py
import json
from collections import deque

def push_event(event_type, payload):
    event = {"type": event_type}
    event.update(payload)
    print(json.dumps(event))

def flood_fill(width, height, start, walls):
    push_event("SYSTEM_LOG", {"message": f"Starting Flood Fill from {start}", "level": "INFO"})
    
    queue = deque([start])
    visited = {start}
    wall_set = set(tuple(w) for w in walls)
    
    if start in wall_set:
        push_event("SYSTEM_LOG", {"message": "Start point is a wall!", "level": "ERROR"})
        return

    # Highlight source
    push_event("MATRIX_CELL_HIGHLIGHT", {"row": start[1], "col": start[0], "color": "#3b82f6"})

    while queue:
        current = queue.popleft()
        
        for dx, dy in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
            neighbor = (current[0] + dx, current[1] + dy)
            
            if 0 <= neighbor[0] < width and 0 <= neighbor[1] < height:
                if neighbor not in visited and neighbor not in wall_set:
                    visited.add(neighbor)
                    queue.append(neighbor)
                    
                    push_event("MATRIX_CELL_HIGHLIGHT", {"row": neighbor[1], "col": neighbor[0], "color": "#0ea5e9"})

    push_event("SYSTEM_LOG", {"message": f"Flood fill completed. Filled {len(visited)} cells.", "level": "INFO"})

if __name__ == "__main__":
    # Default demo grid
    width, height = 20, 20
    start = (10, 10)
    walls = []
    
    flood_fill(width, height, start, walls)

File: src/algorithms/cpp/flood-fill.cpp
/**
 * @file flood-fill.cpp
 * @description C++ implementation of Flood Fill for the sandbox environment.
 * 
 * Time Complexity:  O(V + E)
 * Space Complexity: O(V)
 */

#include <iostream>
#include <vector>
#include <queue>
#include <set>
#include <string>

using namespace std;

/**
 * Sends a visualization event to the backend via stdout.
 */
void push_event(const string& type, const string& payload_json) {
    cout << "{\"type\": \"" << type << "\", " << payload_json << "}" << endl;
}

/**
 * Core Flood Fill Algorithm
 */
void flood_fill(int width, int height, pair<int, int> start, const vector<pair<int, int>>& walls) {
    push_event("SYSTEM_LOG", "\"message\": \"Starting Flood Fill (C++)\", \"level\": \"INFO\"");
    
    set<pair<int, int>> wall_set(walls.begin(), walls.end());
    
    if (wall_set.count(start)) {
        push_event("SYSTEM_LOG", "\"message\": \"Start point is a wall!\", \"level\": \"ERROR\"");
        return;
    }
    
    queue<pair<int, int>> q;
    set<pair<int, int>> visited;
    
    q.push(start);
    visited.insert(start);
    
    // Highlight source
    push_event("MATRIX_CELL_HIGHLIGHT", "\"row\": " + to_string(start.second) + ", \"col\": " + to_string(start.first) + ", \"color\": \"#3b82f6\"");
    
    int filled_count = 0;
    
    while (!q.empty()) {
        pair<int, int> current = q.front();
        q.pop();
        filled_count++;
        
        int dx[] = {0, 1, 0, -1};
        int dy[] = {1, 0, -1, 0};
        
        for (int i = 0; i < 4; i++) {
            pair<int, int> neighbor = {current.first + dx[i], current.second + dy[i]};
            
            if (neighbor.first >= 0 && neighbor.first < width && neighbor.second >= 0 && neighbor.second < height) {
                if (visited.find(neighbor) == visited.end() && wall_set.find(neighbor) == wall_set.end()) {
                    visited.insert(neighbor);
                    q.push(neighbor);
                    
                    push_event("MATRIX_CELL_HIGHLIGHT", "\"row\": " + to_string(neighbor.second) + ", \"col\": " + to_string(neighbor.first) + ", \"color\": \"#0ea5e9\"");
                }
            }
        }
    }
    
    push_event("SYSTEM_LOG", "\"message\": \"Flood fill completed. Filled " + to_string(filled_count) + " cells.\", \"level\": \"INFO\"");
}

int main() {
    int width = 20;
    int height = 20;
    pair<int, int> start = {10, 10};
    vector<pair<int, int>> walls = {};
    
    flood_fill(width, height, start, walls);
    
    return 0;
}
